### PR TITLE
add new fields to generator mapper

### PR DESCRIPTION
**Pull request by YenguiSeddik**

### CODE REVIEW

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java [hunk 1]
```diff
@@ -138,17 +141,57 @@ private static GeneratorMapData toMapData(Generator generator) {
...
+        GeneratorStartup generatorStartup = generator.getExtension(GeneratorStartup.class);
+        if (generatorStartup != null) {
+            builder.marginalCost(generatorStartup.getMarginalCost());
         }
 
         Terminal regulatingTerminal = generator.getRegulatingTerminal();
         //If there is no regulating terminal in file, regulating terminal voltage level is equal to generator voltage level
         if (regulatingTerminal != null && !regulatingTerminal.getVoltageLevel().equals(terminal.getVoltageLevel())) {
             builder.regulatingTerminal(regulatingTerminal.getVoltageLevel().getNameOrId());
+            builder.regulatingTerminalId(regulatingTerminal.getConnectable().getId());
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/network/map/NetworkMapService.java:162]: regulatingTerminalConnectableId ? (a Terminal has no id)

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java [hunk 2]
```diff
@@ -138,17 +141,57 @@ private static GeneratorMapData toMapData(Generator generator) {
...
+            builder.regulatingTerminalType(regulatingTerminal.getConnectable().getType().name());
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/network/map/NetworkMapService.java:163]: regulatingTerminalConnectableType ?

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java [hunk 3]
```diff
@@ -138,17 +141,57 @@ private static GeneratorMapData toMapData(Generator generator) {
...
+
+        GeneratorStartup generatorStartup = generator.getExtension(GeneratorStartup.class);
+        if (generatorStartup != null) {
+            builder.marginalCost(generatorStartup.getMarginalCost());
         }
 
         Terminal regulatingTerminal = generator.getRegulatingTerminal();
         //If there is no regulating terminal in file, regulating terminal voltage level is equal to generator voltage level
         if (regulatingTerminal != null && !regulatingTerminal.getVoltageLevel().equals(terminal.getVoltageLevel())) {
             builder.regulatingTerminal(regulatingTerminal.getVoltageLevel().getNameOrId());
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/network/map/NetworkMapService.java:161]: rename existing field regulatingTerminal into regulatingTerminalVlName ?

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java [hunk 4]
```diff
@@ -138,17 +141,57 @@ private static GeneratorMapData toMapData(Generator generator) {
...
+        }
+
+        ReactiveLimits reactiveLimits = generator.getReactiveLimits();
+        if (reactiveLimits != null) {
+            ReactiveLimitsKind limitsKind = reactiveLimits.getKind();
+            if (limitsKind == ReactiveLimitsKind.MIN_MAX) {
+                MinMaxReactiveLimits minMaxReactiveLimits = generator.getReactiveLimits(MinMaxReactiveLimitsImpl.class);
+                builder.maximumReactivePower(minMaxReactiveLimits.getMaxQ());
+                builder.minimumReactivePower(minMaxReactiveLimits.getMinQ());
+                builder.reactiveCapabilityCurve(false);
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/network/map/NetworkMapService.java:174]: Maybe a class MinMaxReactiveLimitsMapData with minimumReactivePower and maximumReactivePower fields, added in GeneratorMapData ?
So, you could remove reactiveCapabilityCurve booleen field ?

File: src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java [hunk 5]
```diff
@@ -59,4 +61,40 @@ public class GeneratorMapData {
...
+    private String regulatingTerminalId;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private String regulatingTerminalType;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private String regulatingTerminalVlId;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private Boolean reactiveCapabilityCurve;
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java:99]: See remark above

File: src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java [hunk 6]
```diff
@@ -59,4 +61,40 @@ public class GeneratorMapData {
...
     private boolean voltageRegulatorOn;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private Double marginalCost;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private Double minimumReactivePower;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private Double maximumReactivePower;
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java:72]: See remark above

File: src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java [hunk 7]
```diff
@@ -59,4 +61,40 @@ public class GeneratorMapData {
     private Double ratedS;
 
     private boolean voltageRegulatorOn;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private Double marginalCost;
+
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private Double minimumReactivePower;
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java:69]: See remark above

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java (hunk 1 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/network/map/NetworkMapService.java]: done

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java (hunk 3 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/network/map/NetworkMapService.java]: Ok, done

File: src/main/java/org/gridsuite/network/map/NetworkMapService.java (hunk 4 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/network/map/NetworkMapService.java]: Ok, done

File: src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java (hunk 5 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java]: done

File: src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java (hunk 6 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java]: done

File: src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java (hunk 7 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/network/map/model/GeneratorMapData.java]: done

**Review by laurgarn — APPROVED** [review]

File: src/test/resources/generator-map-data.json [hunk 8]
```diff
@@ -2,17 +2,25 @@
...
+  "voltageRegulatorOn":true,
+  "minMaxReactiveLimits": {
+    "minimumReactivePower": -500,
+    "maximumReactivePower": 500
+  },
+  "droop": 4,
+  "regulatingTerminalConnectableId": "NHV1_NHV2_1",
+  "regulatingTerminalConnectableType": "LINE",
+  "regulatingTerminalVlId": "VLHV1"
 }
```
Comment (sBouzols) [inline src/test/resources/generator-map-data.json:26]: add line at end of file

File: src/main/java/org/gridsuite/network/map/model/ReactiveCapabilityCurveMapData.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * @author Seddik Yengui <seddik.yengui at rte-france.com>
+ */
+
+@Builder
+@Getter
+@EqualsAndHashCode
+public class ReactiveCapabilityCurveMapData {
+
+    private Double p;
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/network/map/model/ReactiveCapabilityCurveMapData.java:23]: `@JsonInclude(JsonInclude.Include.NON_NULL)` ?

File: src/test/resources/generators-map-data.json [hunk 10]
```diff
@@ -28,6 +36,23 @@
...
+      },
+      {
+        "p": 3.0,
+        "qminP": 4.0,
+        "qmaxP": 5.0
+      }
+    ]
   }
-]
+]
```
[the remark this answers is not in the document]
Comment (YenguiSeddik) [reply src/test/resources/generators-map-data.json]: Ok. done

File: src/main/java/org/gridsuite/network/map/model/ReactiveCapabilityCurveMapData.java (hunk 9 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/network/map/model/ReactiveCapabilityCurveMapData.java:23]: When reactiveCapabilityCurvePoints of GneratorMapData is no null, the members of ReactiveCapabilityCurveMapData can not be null

**Review by FranckLecuyer — APPROVED** [review]

**Pull request merged**